src/terminal/default.ts

```diff
@@ -0,0 +1,358 @@
+   */
+  private _getName(): Promise<string> {
+    let url = Private.getBaseUrl(this._baseUrl);
+    let ajaxSettings = this.ajaxSettings;
+    ajaxSettings.method = 'POST';
+    ajaxSettings.dataType = 'json';
+
+    return utils.ajaxRequest(url, ajaxSettings).then(success => {
+      if (success.xhr.status !== 200) {
+        return utils.makeAjaxError(success);
+      }
+      return (success.data as TerminalSession.IModel).name;
+    });
+  }
+
+  /**
+   * Connect to the websocket.
+   */
+  private _initializeSocket(): Promise<TerminalSession.ISession> {
+    let name = this._name;
+    this._url = Private.getTermUrl(this._baseUrl, this._name);
+    Private.running[this._url] = this;
+    let wsUrl = utils.urlPathJoin(this._wsUrl, `terminals/websocket/${name}`);
+    this._ws = new WebSocket(wsUrl);
+
+    this._ws.onmessage = (event: MessageEvent) => {
+      let data = JSON.parse(event.data) as JSONPrimitive[];
+      this.messageReceived.emit({
+        type: data[0] as TerminalSession.MessageType,
+        content: data.slice(1)
+      });
+    };
+
+    return new Promise<TerminalSession.ISession>((resolve, reject) => {
+      this._ws.onopen = (event: MessageEvent) => {
+        resolve(this);
+      };
+      this._ws.onerror = (event: Event) => {
+        reject(this);
+      };
+    });
+  }
+
+  private _name: string;
+  private _baseUrl: string;
+  private _wsUrl: string;
+  private _url: string;
+  private _ajaxSettings = '';
+  private _ws: WebSocket = null;
+  private _isDisposed = false;
+  private _promise: Promise<TerminalSession.ISession>;
+}
+
+
+
+export
+namespace DefaultTerminalSession {
+  /**
+   * Start a new terminal session.
+   *
+   * @options - The session options to use.
+   *
+   * @returns A promise that resolves with the session instance.
+   */
+  export
+  function startNew(options: TerminalSession.IOptions = {}): Promise<TerminalSession.ISession> {
+    return new DefaultTerminalSession('', options).connect();
+  }
+
+  /*
+   * Connect to a running session.
+   *
+   * @param name - The name of the target session.
+   *
+   * @param options - The session options to use.
+   *
+   * @returns A promise that resolves with the new session instance.
+   *
+   * #### Notes
+   * If the session was already started via `startNew`, the existing
+   * session object is used as the fulfillment value.
+   *
+   * Otherwise, if `options` are given, we attempt to connect to the existing
+   * session.
+   * The promise is fulfilled when the session is ready on the server,
+   * otherwise the promise is rejected.
+   *
+   * If the session was not already started and no `options` are given,
+   * the promise is rejected.
+   */
+  export
+  function connectTo(name: string, options: TerminalSession.IOptions = {}): Promise<TerminalSession.ISession> {
+    options.baseUrl = options.baseUrl || utils.getBaseUrl();
+    let url = Private.getTermUrl(options.baseUrl, name);
+    if (url in Private.running) {
```

Comment:
You might want to consider using a `Set` for `running`, and checking with the `has()` method since that's a pattern we're starting to use elsewhere.